### ISSUE TITLE
Pass only the property that was changed to the callback

### ISSUE THREAD

**Issue by bramus** [bramus/style-observer#1]
Right now, when any of the properties change, the full set of tracked properties gets passed into the callback. This can change since `TransitionEvent` has a `propertyName` property which can be used to determine which property to pass to the callback.

This could be put behind a flag, where authors can opt-in to choose between the blanket behavior, or the selective behavior.

**Comment by bramus** [comment]
An alternative would be to change the shape of the returned results. It would still be an array but each result would be an object.

Something like this:

```
{
  propertyName: string,
  value: any,
  changed: boolean,
}
```

**Comment by faizanu94** [comment]
I’ve submitted a [PR](https://github.com/bramus/style-observer/pull/3) that implements the selective mode as initially suggested. This mode passes only the changed property to the callback. The implementation triggers a callback for each property change, ensuring immediate responsiveness to each change event without delays

**Comment by bramus** [comment]
Thanks for doing this PR. While the PR definitely is in a good state now after us going back and forth on it, I noticed it won't actually work

The culprit is this clause in the spec:

> if one of the identifiers listed is a shorthand property, implementations must start transitions for all its longhand sub-properties

For `border-width` for example, you get 4 transition events: `border-top-width`, `border-right-width`, `border-bottom-width` and `border-left-width`.

I think we need to either come up with a solution (and preferably one that doesn't require specifying all longhands and shorthands) or abandon the original idea.

Part of the possible solution laid out in comment 2 here might work: cache the values, diff ourselves, and only dispatch the callback with only the changed values.

**Comment by faizanu94** [comment]
After reviewing your feedback and considering the CSS spec's implications about shorthand properties triggering transitions for all longhand sub-properties, I agree that the idea mentioned in comment 2 seems like a better approach to move forward!

Would implementing a structure like below for the returned results be a good way to go? This change would keep the package's API same while providing detailed information about each property:

```json
[
  {"propertyName": "margin", "value": "40px", "changed": true},
  {"propertyName": "color", "value": "rgb(0, 0, 0)", "changed": false}
]
```

As for expanding shorthand properties to their longhand equivalents, I have considered a method to do so, but I agree that it feels like a workaround. Here’s a basic version of what that could look like:

```typescript
const _expandShorthandProperties = (property: string) => {
  const shorthandMapping: Record<string, string[]> = {
    margin: ['margin-top', 'margin-right', 'margin-bottom', 'margin-left'],
    padding: ['padding-top', 'padding-right', 'padding-bottom', 'padding-left'],
    border: ['border-top', 'border-right', 'border-bottom', 'border-left'],
    'border-width': ['border-top-width', 'border-right-width', 'border-bottom-width', 'border-left-width'],
    'border-color': ['border-top-color', 'border-right-color', 'border-bottom-color', 'border-left-color'],
    'border-style': ['border-top-style', 'border-right-style', 'border-bottom-style', 'border-left-style'],
  };
  return shorthandMapping[property] || [property];
};
```

Depending on the consensus, we can either modify the existing PR or start a new one

**Comment by bramus** [comment]
> Would implementing a structure like below for the returned results be a good way to go? This change would keep the package's API same while providing detailed information about each property:
> 
> ```json
> [
>   {"propertyName": "margin", "value": "40px", "changed": true},
>   {"propertyName": "color", "value": "rgb(0, 0, 0)", "changed": false}
> ]
> ```

Just a thought: If we’re keeping track of the before value internally, we might as well include that for each result. That way authors can know what the value was before and what it is now.

---

> As for expanding shorthand properties to their longhand equivalents, I have considered a method to do so, but I agree that it feels like a workaround. Here’s a basic version of what that could look like:

I noticed it is a bit more complicated then it is right now, as changing `border` directly does not trigger the listed `border-width`/`border-color`/`border-style`. Changing `border` from `1px solid black` to `10px solid black` does not transition `border-width` but transitions its longhands `border-top-width` + `border-right-width` + `border-bottom-width` and `border-left-width` – there needs to be some recursive expansion.

That said, the final list will be way longer than just those properties, as there are many more. Thankfully such a list could be generated from https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/css_properties.json5

Checking `border-width` for example, you can see that the longhands are listed there:

```
    {
      name: "border-width",
      longhands: [
        "border-top-width", "border-right-width", "border-bottom-width",
        "border-left-width"
      ],
      property_methods: ["ParseShorthand", "CSSValueFromComputedStyleInternal"],
      valid_for_page_context: true,
    },
```

Some things to consider here:

- Having such a list could increase the size of the package significantly, which is something that needs to be measured.
- That list is Chromium only. If a different UA supports a property that is not supported by Blink (yet), then the it would not be included in the generated list.
- There would also need to be some mapping from logical properties to the non-logical ones

But say we can somehow come up with a complete list. In that case there would still need to be some local caching of the computed value, so that style-observer can selectively trigger the callback.

For example, when monitoring `border` and then changing the `border-width`, all 4 `border-top-width` + `border-right-width` + `border-bottom-width` and `border-left-width` transitions get triggered. When the first event comes in, style-observer needs to compute `border` to decide whether it should invoke `callback` or not, after which it should re-cache the value so that the successive 3 triggers don’t trigger callback as well … at which point I think we are back at the same place as the alternative implementation – where a transition simply gets all computed values for all observed properties – provides us.

tl;dr I am currently not convinced the shorthand-longhand expansion is worth the effort. AFAICT it doesn’t add any real benefits.

**Comment by faizanu94** [comment]
Thanks for the feedback - I agree that expanding shorthand properties to their longhand equivalents would add complexity without significant benefit. Instead, as suggested in comment 2, I’ve implemented a [solution](https://github.com/bramus/style-observer/pull/6) that caches values and checks for differences internally. This way, we can efficiently detect changes and trigger the callback only when necessary

The updated response structure includes `propertyName`, `value`, `previousValue` and `changed` providing more detailed information about each observed property. This approach keeps the package's API unchanged while enhancing its functionality

**Comment by bramus** [comment]
With the merge of #7, the original request from this issue was fixed. I have filed https://github.com/bramus/style-observer/issues/8 to track the extended return format which was also discussed in this thread.
